core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -339,64 +350,74 @@ class Partition(val topicPartition: TopicPartition,
-          } else {
-            Left(replica)
+          log match {
+            case Some(partitionLog) =>
+              Left(partitionLog)
+            case _ =>
+              if (requireLeader)
+                Right(Errors.NOT_LEADER_FOR_PARTITION)
+              else
+                Right(Errors.REPLICA_NOT_AVAILABLE)
           }
         }
       case error =>
         Right(error)
     }
   }
 
-  def localReplica: Option[Replica] = getReplica(localBrokerId)
-
-  def localReplicaOrException: Replica = localReplica.getOrElse {
-    throw new ReplicaNotAvailableException(s"Replica for partition $topicPartition is not available " +
+  def localLogOrException: Log = log.getOrElse {
+    throw new ReplicaNotAvailableException(s"Log for partition $topicPartition is not available " +
       s"on broker $localBrokerId")
   }
 
-  def futureLocalReplica: Option[Replica] = getReplica(Request.FutureLocalReplicaId)
+  private def localReplica: Option[LocalReplica] = getReplica(localBrokerId).asInstanceOf[Option[LocalReplica]]
 
-  def futureLocalReplicaOrException: Replica = futureLocalReplica.getOrElse {
-    throw new ReplicaNotAvailableException(s"Future replica for partition $topicPartition is not available " +
+  def futureLocalLogOrException: Log = futureLog.getOrElse {
+    throw new ReplicaNotAvailableException(s"Future log for partition $topicPartition is not available " +
       s"on broker $localBrokerId")
   }
 
-  def leaderReplicaIfLocal: Option[Replica] = {
-    if (leaderReplicaIdOpt.contains(localBrokerId))
-      localReplica
-    else
-      None
+  def leaderLogIfLocal: Option[Log] = {
+    log.filter(_ => isLeader)
   }
 
-  private def localReplicaWithEpochOrException(currentLeaderEpoch: Optional[Integer],
-                                               requireLeader: Boolean): Replica = {
-    getLocalReplica(localBrokerId, currentLeaderEpoch, requireLeader) match {
-      case Left(replica) => replica
+  /**
+   * Returns true if this node is currently leader for the Partition.
+   */
+  def isLeader: Boolean = leaderReplicaIdOpt.contains(localBrokerId)
+
+  private def localLogWithEpochOrException(currentLeaderEpoch: Optional[Integer],
+                                           requireLeader: Boolean): Log = {
+    getLocalLog(currentLeaderEpoch, requireLeader) match {
+      case Left(localLog) => localLog
       case Right(error) =>
         throw error.exception(s"Failed to find ${if (requireLeader) "leader " else ""} for " +
           s"partition $topicPartition with leader epoch $currentLeaderEpoch. The current leader " +
           s"is $leaderReplicaIdOpt and the current epoch $leaderEpoch")
     }
   }
 
-  // Visible for testing
-  def addReplicaIfNotExists(replica: Replica): Replica =
+  // Visible for testing -- Only used in tests to add replica to existing partition
+  def addReplicaIfNotExists(replica: Replica): Replica = {
     allReplicasMap.putIfNotExists(replica.brokerId, replica)
+    replica match {
```

Comment:
Done.

This method is only used by test code, I think we should enforce this somehow as we don't want real code to start using it. When Partition is created, its log and replica is assigned as part of standard state transition. Test code however wants to create a replica/log object and partition object separately (mostly mocks) and then tie them together using this method. I think its ok to do this part of test, but it would be nice to not let non-test code use it. In c++ world friend classes can be used for this purpose, maybe in Scala we have similar pattern which I am not aware of.

Reply:
Agreed. I think we can probably refactor the tests to always go through `makeLeader`, `makeFollower`, or `maybeCreateFutureReplica`. Then we wouldn't need this anymore. I did this in a few localized cases already.

Reply:
> Scala we have similar pattern which I am not aware of

Not sure if it applies here but you can make it package accessible only as long as your test are also in the same package. E.g. `private[cluster] ...`

My understanding is that as long as you don't use `private[this]` the JVM code generated is `public` and it is only the scala compiler that is enforcing visibility.

Reply:
Used outside package. Left is as is, relying on method comment and review to enforce the encapsulation.